core/src/main/scala/org/ensime/core/Indexer.scala

```diff
@@ -73,16 +73,26 @@ class Indexer(
       import context.dispatcher
       val usages = index.findUsageLocations(fqn)
       val response = usages.map { usages =>
-        val results: List[LineSourcePosition] = usages.flatMap { u =>
+        val positions: List[LineSourcePosition] = usages.take(1000).flatMap { u =>
           val file = u.file
           file.map { f =>
             LineSourcePosition(
               EnsimeFile(Paths.get(new URI(f)).toString),
-              u.line.getOrElse(u.line.getOrElse(0))
+              u.line.getOrElse(0)
             )
           }
         }(collection.breakOut)
-        SourcePositions(results)
+        val files = positions.foldLeft[Set[EnsimeFile]](Set.empty) { (files, pos) =>
+          if (pos.line > 0) files + pos.file else files
+        }
+        val contents: Map[EnsimeFile, Array[String]] = files.map(f => f -> f.readAllLines.toArray)(collection.breakOut)
```

Comment:
oh, I missed the `toArray` here, it is almost certainly slower to create an `Array` and lookup than the cost to lookup the line directly on the `List`.